### PR TITLE
feat (jkube-kit/resource/helm) : Support for Pushing Helm Charts to OCI registry (#2091)

### LINKED ISSUE #1364: Support pushing of Helm charts to OCI registries
### Description
With the release of Helm 3.8.0, Helm is able to store and work with charts in container registries, as an alternative to Helm repositories (c.f. [Storing Helm Charts in OCI Registries](https://helm.sh/blog/storing-charts-in-oci/)).

This is a feature request to add support for oci in the `helm-push` goal.

### Info

* Eclipse JKube version : 1.7.0
* Maven version (`mvn -v`) :
```
Maven home: /usr/local/Cellar/maven/3.8.3/libexec
Java version: 17, vendor: Homebrew, runtime: /usr/local/Cellar/openjdk/17/libexec/openjdk.jdk/Contents/Home
Default locale: en_FR, platform encoding: US-ASCII
OS name: "mac os x", version: "12.2.1", arch: "x86_64", family: "mac"
```

This would provide a vendor independant alternative to the existing HTTP based repositories currently supported.

### LINKED ISSUE #2091: Helm: Support pushing to OCI registry
> With the release of Helm 3.8.0, Helm is able to store and work with charts in container registries, as an alternative to [Helm repositories](https://helm.sh/docs/topics/chart_repository/). This feature, which used to be an experimental feature, is now generally available.

https://helm.sh/blog/storing-charts-in-oci/

Providing support for this should allow pushing charts to (for example):
- Quay.io
- Docker Hub
- GitHub packages

### PR DESCRIPTION
**Pull request by rohanKanojia**
## Description
Fixes #2091
Fixes #1364

Support for Pushing Helm Charts to OCI registry

+ Add SHAUtil for generating SHA digests (used in oci helm push)
+ Add OCI repository type to HelmRepoType enum
+ Add OCIRepositoryUploader as an additional HelmUploader
+ Add OCI repository-related helper classes in `org.eclipse.jkube.kit.resource.helm.oci` package
   - OCIManifest : POJO for chart upload metadata 
   - OCIManifestLayer: POJO for chart upload metadata (used by OCIManifest)
   - OCIRegistryAuthenticator: Responsible for getting OAuth token before uploading helm chart
   - OCIRegistryClient: contains various utility methods as steps for OCI Helm Upload

Signed-off-by: Rohan Kumar <rohaan@redhat.com>

## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [X] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [X] I have read the [contributing guidelines](https://www.eclipse.org/jkube/contributing)
 - [X] I signed-off my commit with a user that has signed the [Eclipse Contributor Agreement](https://www.eclipse.org/legal/ECA.php)
 - [X] My code follows the style guidelines of this project
 - [ ] I Keep It Small and Simple: The smaller the PR is, the easier it is to review and have it merged
 - [X] I use [conventional commits](https://www.conventionalcommits.org/) in my commit messages
 - [ ] I have performed a self-review of my code
 - [X] I Added [CHANGELOG](../CHANGELOG.md) entry
 - [X] I have updated the [documentation](../kubernetes-maven-plugin/doc) accordingly
 - [X] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=jkubeio_jkube) report
 - [X] I have added tests that prove my fix is effective or that my feature works
 - [X] New and existing unit tests pass locally with my changes
 - [X] I tested my code in Kubernetes
 - [X] I tested my code in OpenShift

### CODE REVIEW

**Comment by manusa** [comment]
We've created yet another HTTP client implementation in this PR.

I'm not sure if this was discussed prior to implementing the solution, but it's certainly a debatable topic. Especially because we already have at least 2 HTTP clients in the project. We should be working towards keeping a single one, not adding more to the list.

File: jkube-kit/parent/pom.xml [hunk 1]
```diff
@@ -677,6 +677,14 @@
       </dependency>
       <!-- == test ====================================== -->
 
+      <dependency>
+        <groupId>io.fabric8</groupId>
+        <artifactId>kubernetes-client-api</artifactId>
+        <type>test-jar</type>
+        <version>${version.kubernetes-client}</version>
+        <scope>test</scope>
+      </dependency>
```
Comment (manusa) [inline jkube-kit/parent/pom.xml:686]: Considering the Fabric8 BOM is added as a dependencyManagement entry? is this entry necessary?

File: jkube-kit/resource/helm/src/main/java/org/eclipse/jkube/kit/resource/helm/HelmService.java [hunk 2]
```diff
@@ -160,7 +160,7 @@ private void uploadHelmChart(HelmConfig helmConfig, HelmRepository helmRepositor
       throws IOException, BadUploadException {
 
     for (HelmConfig.HelmType helmType : helmConfig.getTypes()) {
-      final HelmUploaderManager helmUploaderManager = new HelmUploaderManager(logger);
+      final HelmUploaderManager helmUploaderManager = new HelmUploaderManager();
```
Comment (manusa) [inline jkube-kit/resource/helm/src/main/java/org/eclipse/jkube/kit/resource/helm/HelmService.java:163]: This should be initiated outside of the for-loop

File: jkube-kit/resource/helm/src/main/java/org/eclipse/jkube/kit/resource/helm/OCIRepositoryUploader.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+  @Override
+  public HelmRepository.HelmRepoType getType() {
+    return HelmRepository.HelmRepoType.OCI;
+  }
+
+  @Override
+  public void uploadSingle(KitLogger logger, File file, HelmRepository repository)
+      throws IOException, BadUploadException {
+    Chart chartConfig = createChartMetadataFromGeneratedChartYamlFile(file);
+    try (HttpClient httpClient = HttpClientUtils.getHttpClientFactory().newBuilder().tag(new RequestConfigBuilder().withRequestRetryBackoffLimit(0).build()).build()) {
```
Comment (manusa) [inline jkube-kit/resource/helm/src/main/java/org/eclipse/jkube/kit/resource/helm/OCIRepositoryUploader.java:42]: Why are we disabling the retry backoff limit?

File: jkube-kit/parent/pom.xml (hunk 1 above)
Comment (rohanKanojia) [reply jkube-kit/parent/pom.xml:686]: project fails to compile if I remove this entry:
```
 [INFO] Scanning for projects...
[ERROR] [ERROR] Some problems were encountered while processing the POMs:
[ERROR] 'dependencies.dependency.version' for io.fabric8:kubernetes-client-api:test-jar is missing. @ line 133, column 17
[ERROR] 'dependencies.dependency.version' for io.fabric8:kubernetes-client-api:test-jar is missing. @ line 52, column 17
 @ 
[ERROR] The build could not read 2 projects -> [Help 1]
[ERROR]   
[ERROR]   The project org.eclipse.jkube:jkube-kit-common:1.14-SNAPSHOT (/home/rokumar/work/repos/jkube/jkube-kit/common/pom.xml) has 1 error
[ERROR]     'dependencies.dependency.version' for io.fabric8:kubernetes-client-api:test-jar is missing. @ line 133, column 17
[ERROR]   
[ERROR]   The project org.eclipse.jkube:jkube-kit-resource-helm:1.14-SNAPSHOT (/home/rokumar/work/repos/jkube/jkube-kit/resource/helm/pom.xml) has 1 error
[ERROR]     'dependencies.dependency.version' for io.fabric8:kubernetes-client-api:test-jar is missing. @ line 52, column 17
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/ProjectBuildingException

```

File: jkube-kit/resource/helm/src/main/java/org/eclipse/jkube/kit/resource/helm/HelmUploader.java [hunk 4]
```diff
@@ -13,10 +13,12 @@
...
 
+import org.eclipse.jkube.kit.common.KitLogger;
+
 import java.io.File;
 import java.io.IOException;
 
 public interface HelmUploader {
   HelmRepository.HelmRepoType getType();
-  void uploadSingle(File file, HelmRepository repository) throws IOException, BadUploadException ;
+  void uploadSingle(KitLogger kitLogger, File file, HelmRepository repository) throws IOException, BadUploadException ;
```
Comment (manusa) [inline jkube-kit/resource/helm/src/main/java/org/eclipse/jkube/kit/resource/helm/HelmUploader.java:23]: Considering the logger is used to publish low-level messages (which might not be informative to end users), maybe we want to reconsider passing in the logger at all.

File: jkube-kit/resource/helm/src/main/java/org/eclipse/jkube/kit/resource/helm/oci/OCIRegistryAuthenticator.java [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.eclipse.jkube.kit.common.util.AsyncUtil.get;
+import static org.eclipse.jkube.kit.common.util.Fabric8HttpUtil.extractAuthenticationChallengeIntoMap;
+import static org.eclipse.jkube.kit.common.util.Fabric8HttpUtil.toFormData;
+
+
+public class OCIRegistryAuthenticator {
```
Comment (manusa) [inline jkube-kit/resource/helm/src/main/java/org/eclipse/jkube/kit/resource/helm/oci/OCIRegistryAuthenticator.java:37]: If OCI is authenticated **only** on failure (401), then maybe we want to implement this as an interceptor

**Comment by manusa** [comment]
All tests are currently relying on Mocks.
Probably most of them can be refactored into a few that rely on a real HTTP conversation.
I'd also like to see the new implementation being tested in HelmServiceUploadIT.

**Review by manusa — CHANGES_REQUESTED** [review]
see previous comments

File: jkube-kit/common/src/main/java/org/eclipse/jkube/kit/common/util/SHAUtil.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+
+public class SHAUtil {
```
Comment (manusa) [inline jkube-kit/common/src/main/java/org/eclipse/jkube/kit/common/util/SHAUtil.java:24]: Most of the utils in this class are already available in the [`AwsSigner4`](https://github.com/eclipse/jkube/blob/03554382b7c292ebffd4133388805fe39cac0ccd/jkube-kit/build/service/docker/src/main/java/org/eclipse/jkube/kit/build/service/docker/auth/ecr/AwsSigner4.java)

We might want to consolidate those prior to continue the work on this PR.

Also, we're reimplementing a read file approach which I'm sure we have elsewehere in the repo too.

Comment (manusa) [reply jkube-kit/common/src/main/java/org/eclipse/jkube/kit/common/util/SHAUtil.java:24]: In general the idea would be to reuse whatever other **core** libraries we rely on to delegate these simple/common tasks

Comment (rohanKanojia) [reply jkube-kit/common/src/main/java/org/eclipse/jkube/kit/common/util/SHAUtil.java:24]: You're right. I was able to replace both methods using [Apache commons codec's DigestUtils](https://commons.apache.org/proper/commons-codec/apidocs/org/apache/commons/codec/digest/DigestUtils.html)

**Review by manusa — APPROVED** [review]

**Pull request merged**